app/services/topics/mutator.rb

```diff
@@ -0,0 +1,146 @@
+class Topics::Mutator
+  def initialize(topic:, params: nil, document_signed_ids: nil)
+    @topic = topic
+    @params = params
+    @document_signed_ids = document_signed_ids
+  end
+
+  def create
+    mutate
+  end
+
+  def update
+    mutate
+  end
+
+  def archive
+    if @topic.archived!
+      sync_docs_for_topic_archive if topic.documents.any?
+      return [ :ok, topic ]
+    end
+
+    [ :error, topic.errors.full_messages ]
+  end
+
+  def unarchive
+    if @topic.active!
+      sync_docs_for_topic_archive if topic.documents.any?
+      return [ :ok, topic ]
+    end
+
+    [ :error, topic.errors.full_messages ]
+  end
+
+  def destroy
+    # If topic deletion fails for some reason, documents deletion still will happen
+    # This case is unlikely, and only admins can delete topics
+    sync_docs_for_topic_deletion if topic.documents.any?
+    return [ :ok, topic ] if topic.destroy
+
+    [ :error, topic.errors.full_messages ]
+  end
+
+  private
+
+  attr_reader :topic, :params, :document_signed_ids
+
+  def mutate
+    topic.valid?
+    return [ :error,  topic.errors.full_messages ] if topic.errors.any?
+
+    docs_to_delete = outdated_documents
```

Comment:
In some documents are no longer belong to topic (for example, after update) we still need to handle them properly (remove from storage). This is where we will use "shadow copy"